### PR TITLE
feat: Build CAC Lang using a formal expression parser

### PR DESCRIPTION
**Pull request by knutties**
## Problem
This PR attempts to improve the existing caclang implementation by using a formal parser and formalize the file format for cac. 

## Solution
The solution leverages the [pest parser](https://pest.rs/) to parse context expressions and continues to use the toml format for cac file.  It also captures more detail for the default config key around types.  It is a start towards maintaining feature parity between the file based cac implementation and the web-service.

The PR also updates the documentation to use the exact same format for the examples as is used by the code.  It also includes minor updates to the base README.md including details of twitter/discord presence.

## Environment variable changes
NA

## Pre-deployment activity
NA

## Post-deployment activity
NA

## API changes
| Endpoint        | Method           | Request body  | Response Body |
| ------------- |:-------------:| -----:| ----------------:|
NA

## Possible Issues in the future
NA

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,3 +34,9 @@ Once you run this command, you'll find the demo app at `localhost:8081` and Supe
 * **Type/Validation support** - Comprehensive type support using json-schema and custom validator function support for configuration values
 * **Multi-tenant support** - a single deployment allows multiple tenants to manage their configurations/experiments in a completely isolated manner
 * **Authn/Authz support** - control who can make configuration/experimentation changes
+
+## Interested to know more - reach out to the team !
+* Website: https://juspay.io/superposition
+* Twitter: https://x.com/superpositionjp
+* Discord: https://discord.gg/sBgYYUQU
+* Github Issues: https://github.com/juspay/superposition/issues
```
Comment (Datron) [inline README.md:42]: @knutties we can use badges for these?
twitter:
https://img.shields.io/badge/X-000000?style=for-the-badge&logo=x&logoColor=white
discord:
https://img.shields.io/badge/Discord-5865F2?style=for-the-badge&logo=discord&logoColor=white
github:
https://img.shields.io/badge/GitHub-100000?style=for-the-badge&logo=github&logoColor=white

File: crates/cac_toml/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+}
+
+impl PartialEq for ContextualOverride {
+    fn eq(&self, other: &Self) -> bool {
+        (self.extracted_dimensions == other.extracted_dimensions)
+            && (self.expression == other.expression)
+    }
+}
+
+impl Eq for ContextualOverride {}
```
Comment (Datron) [inline crates/cac_toml/src/lib.rs:51]: nitpick: We can derive Eq on the struct itself

File: crates/cac_toml/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+}
+
+#[derive(Debug, Clone)]
+pub struct CACParseError;
+
+impl fmt::Display for CACParseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Unable to parse CAC TOML file.")
+    }
+}
```
Comment (Datron) [inline crates/cac_toml/src/lib.rs:80]: nitpick: lets use strum here in a later PR

File: docs/cac-toml.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
+# Toml formatted CAC configuration file in a simple rust application
+This document shows how to use a TOML formatted CAC configuration file in your Rust application.
+
+## Add dependencies to your Cargo.toml
+
+```toml
+[dependencies]
+cac_toml = { git = "https://github.com/juspay/superposition.git", branch = "pest-cac-lang" }
+toml = { version = "0.8.8", features = ["preserve_order"] }
```
Comment (Datron) [inline docs/cac-toml.md:9]: we should remove the branch here, once merged to main this branch won't exist

File: examples/cac_toml_app_example/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "cac_toml_app_example"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+cac_toml = { git = "https://github.com/juspay/superposition.git", branch = "pest-cac-lang" }
```
Comment (Datron) [inline examples/cac_toml_app_example/Cargo.toml:9]: Can we remove branch here?

**Review by Datron — CHANGES_REQUESTED** [review]
@knutties you have to make changes in cog.toml too and change caclang to cactoml so that versioning happens

File: README.md (hunk 1 above)
Comment (knutties) [reply README.md:42]: @Datron - Just replace the labels with icons or even the links themselves ?  Or do we keep the icon and the label - atleast for X twitter seems still needed no ?

File: docs/cac-toml.md (hunk 4 above)
Comment (knutties) [reply docs/cac-toml.md]: The default branch if omitted is `master` - so will make this `main` for now.

File: crates/cac_toml/src/lib.rs (hunk 2 above)
Comment (knutties) [reply crates/cac_toml/src/lib.rs:51]: @Datron - did not understand, newbie question - you mean using a macro?

Comment (knutties) [reply crates/cac_toml/src/lib.rs:51]: Screenshot 2024-09-06 at 3 06 10 PM

I think I will need to define Eq for Value as well (which is a type I don't own)

**Review by Datron — APPROVED** [review]

**Review by mahatoankitkumar — APPROVED** [review]

**Pull request merged**